docs/content/guides/dagster/pythonic-resources.mdx

```diff
@@ -0,0 +1,440 @@
+```
+
+### Using bare Python objects as resources
+
+In some cases, you may want to use a bare Python object as a resource which is not a subclass of <PyObject object="ConfigurableResource"/>.
+
+For example, you may want to directly pass a third-party API client into your assets or ops. This follows a similar pattern to using a <PyObject object="ConfigurableResource"/> subclass, however assets and ops which use these resources must [annotate](https://docs.python.org/3/library/typing.html#typing.Annotated) them with <PyObject object="Resource"/>.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_raw_github_resource endbefore=end_raw_github_resource
+@asset
+def public_github_repos(github: Resource[GitHub]):
+    return github.organization("dagster-io").repositories()
+
+
+defs = Definitions(
+    assets=[public_github_repos],
+    resources={"github": GitHub()},
+)
+```
+
+If you would like a Pythonic resource to depend on a bare Python object, you can use the <PyObject object="ResourceDependency"/> annotation to annotate the attribute in question. This indicates to Dagster that the attribute should be treated as a resource dependency.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_raw_github_resource_dep endbefore=end_raw_github_resource_dep
+class DBResource(ConfigurableResource):
+    engine: ResourceDependency[Engine]
+
+    def query(self, query: str):
+        with self.engine.connect() as conn:
+            return conn.execute(query)
+
+
+engine = create_engine(...)
+defs = Definitions(
+    assets=...,
+    resources={"db": DBResource(engine=engine)},
+)
+```
+
+---
+
+## Implementation details
+
+### Resource input resolution
+
+Resources defined by subclassing <PyObject object="ConfigurableResource"/> can take various unresolved inputs, including `EnvVar`s instead of string values or unconfigured resources returned by `configure_at_launch()` in place of configured resources.
+
+Between construction time and runtime, the <PyObject object="ConfigurableResource"/> class resolves all of its attributes to their final values, meaning that when an attribute of a resource is accessed in the body of an asset or op, it will always be a fully resolved value.
+
+In the following example, the `FileStoreBucket` constructor will accept the still-unconfigured `unconfigured_credentials_resource`, which will be resolved once the resource is configured at runtime.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_impl_details_resolve endbefore=end_impl_details_resolve
+class CredentialsResource(ConfigurableResource):
+    username: str
+    password: str
+
+
+class FileStoreBucket(ConfigurableResource):
+    credentials: CredentialsResource
+    region: str
+
+    def write(self, data: str):
+        # In this context, `self.credentials` is ensured to
```

Comment:
"In this context, `self.credentials` is ensured to be a CredentialsResource with valid values for `username` and `password`"